llama_index/vector_stores/neo4jvector.py

```diff
@@ -0,0 +1,289 @@
+from collections import namedtuple
+from typing import Any, Dict, List, Optional, Type
+
+from llama_index.schema import MetadataMode, TextNode
+from llama_index.vector_stores.types import (
+    MetadataFilters,
+    NodeWithEmbedding,
+    VectorStore,
+    VectorStoreQuery,
+    VectorStoreQueryResult,
+)
+from llama_index.vector_stores.utils import metadata_dict_to_node, node_to_metadata_dict
+
+
+def check_if_not_null(props: List[str], values: List[Any]) -> None:
+    """Check if variable is not null and raise error accordingly"""
+    for prop, value in zip(props, values):
+        if not value:
+            raise ValueError(f"Parameter `{prop}` must not be None or empty string")
+
+
+def sort_by_index_name(
+    lst: List[Dict[str, Any]], index_name: str
+) -> List[Dict[str, Any]]:
+    """Sort first element to match the index_name if exists"""
+    return sorted(lst, key=lambda x: x.get("index_name") != index_name)
+
+
+def clean_params(params: List[NodeWithEmbedding]) -> List[Dict[str, Any]]:
+    """Convert NodeWithEmbedding object to a dictionary ready to be imported into Neo4j"""
+    clean_params = []
+    for record in params:
+        text = record.node.get_content(metadata_mode=MetadataMode.NONE)
+        embedding = record.embedding
+        id = record.node.ref_doc_id
+        metadata = node_to_metadata_dict(
+            record.node, remove_text=True, flat_metadata=False
+        )
+        # Remove redundant metadata information
+        for k in ["document_id", "ref_doc_id", "doc_id"]:
+            del metadata[k]
+        clean_params.append(
+            {"text": text, "embedding": embedding, "id": id, "metadata": metadata}
+        )
+    return clean_params
+
+
+class Neo4jVectorStore(VectorStore):
+    stores_text: bool = True
+    flat_metadata = True
+
+    def __init__(
+        self,
+        username: str,
+        password: str,
+        url: str,
+        embedding_dimension: int,
+        *,
+        database: str = "neo4j",
+        index_name: str = "vector",
+        node_label: str = "Chunk",
+        embedding_node_property: str = "embedding",
+        text_node_property: str = "text",
+        distance_strategy: str = "cosine",
+        retrieval_query: str = "",
```

Comment:
Just curious, why would I want to hardcode a retrieval string in the constructor?

Reply:
You can add `MATCH` clauses and fetch or calculate additional graph data